Source/DafnyStandardLibraries/src/DafnyStdLibs/EnableNonLinearArithmetic/JSON/Deserializer.dfy

```diff
@@ -0,0 +1,180 @@
+        'A' as uint16 := 0xA, 'B' as uint16 := 0xB, 'C' as uint16 := 0xC, 'D' as uint16 := 0xD, 'E' as uint16 := 0xE, 'F' as uint16 := 0xF
+      ]
+  }
+
+  const HEX_TABLE_16 := Uint16StrConversion.charToDigit
+
+  function ToNat16(str: Uint16StrConversion.String): uint16
+    requires |str| <= 4
+    requires forall c | c in str :: c in HEX_TABLE_16
+  {
+    assume {:axiom} false; // BUG Verification inconclusive
+    Uint16StrConversion.ToNatBound(str);
+    var hd := Uint16StrConversion.ToNat(str);
+    assert hd < 0x1_0000 by { reveal Pow(); }
+    hd as uint16
+  }
+
+  function {:tailrecursion} {:vcs_split_on_every_assert} Unescape(str: seq<uint16>, start: nat := 0, prefix: seq<uint16> := []): DeserializationResult<seq<uint16>>
+    decreases |str| - start
+  { // Assumes UTF-16 strings
+    if start >= |str| then Success(prefix)
+    else if str[start] == '\\' as uint16 then
+      if |str| == start + 1 then
+        Failure(EscapeAtEOS)
+      else
+        var c := str[start + 1];
+        if c == 'u' as uint16 then
+          if |str| <= start + 6 then
+            Failure(EscapeAtEOS)
+          else
+            var code := str[start + 2..start + 6];
+            if exists c | c in code :: c !in HEX_TABLE_16 then
+              Failure(UnsupportedEscape16(code))
+            else
+              var hd := ToNat16(code);
+              Unescape(str, start + 6, prefix + [hd])
+        else
+          var unescaped: uint16 := match c
+            case 0x22 => 0x22 as uint16 // \" => quotation mark
+            case 0x5C => 0x5C as uint16 // \\ => reverse solidus
+            case 0x62 => 0x08 as uint16 // \b => backspace
+            case 0x66 => 0x0C as uint16 // \f => form feed
+            case 0x6E => 0x0A as uint16 // \n => line feed
+            case 0x72 => 0x0D as uint16 // \r => carriage return
+            case 0x74 => 0x09 as uint16 // \t => tab
+            case _    => 0 as uint16;
+          if unescaped as int == 0 then
+            Failure(UnsupportedEscape16(str[start..start+2]))
+          else
+            Unescape(str, start + 2, prefix + [unescaped])
+    else
+      Unescape(str, start + 1, prefix + [str[start]])
+  }
+
+  function String(js: Grammar.jstring): DeserializationResult<string> {
+    var asUtf32 :- FromUTF8Checked(js.contents.Bytes()).ToResult(DeserializationError.InvalidUnicode);
+    var asUint16 :- ToUTF16Checked(asUtf32).ToResult(DeserializationError.InvalidUnicode);
+    var unescaped :- Unescape(asUint16);
+    FromUTF16Checked(unescaped).ToResult(DeserializationError.InvalidUnicode)
+  }
+
+  module ByteStrConversion refines Strings.ParametricConversion {
+    import opened BoundedInts
+
+    type Char = byte
+
+    const chars := [
+      '0' as uint8, '1' as uint8, '2' as uint8, '3' as uint8,
+      '4' as uint8, '5' as uint8, '6' as uint8, '7' as uint8,
+      '8' as uint8, '9' as uint8
+    ]
+
+    const charToDigit := map[
```

Comment:
This probably belongs in the Strings library now instead, but fine to move it there later